cpp/src/parquet/size_statistics_test.cc

```diff
@@ -0,0 +1,307 @@
+    ASSERT_OK_AND_ASSIGN(buffer_, sink->Finish());
+  }
+
+  void ReadSizeStatistics() {
+    auto read_properties = default_arrow_reader_properties();
+    auto reader =
+        ParquetFileReader::Open(std::make_shared<::arrow::io::BufferReader>(buffer_));
+
+    // Read row group size statistics in order.
+    auto metadata = reader->metadata();
+    for (int i = 0; i < metadata->num_row_groups(); ++i) {
+      auto row_group_metadata = metadata->RowGroup(i);
+      for (int j = 0; j < metadata->num_columns(); ++j) {
+        auto column_metadata = row_group_metadata->ColumnChunk(j);
+        auto size_stats = column_metadata->size_statistics();
+        SizeStatistics row_group_stats;
+        if (size_stats != nullptr) {
+          row_group_stats = {size_stats->repetition_level_histogram,
+                             size_stats->definition_level_histogram,
+                             size_stats->unencoded_byte_array_data_bytes};
+        }
+        row_group_stats_.emplace_back(std::move(row_group_stats));
+      }
+    }
+
+    // Read page size statistics in order.
+    auto page_index_reader = reader->GetPageIndexReader();
+    ASSERT_NE(page_index_reader, nullptr);
+
+    for (int i = 0; i < metadata->num_row_groups(); ++i) {
+      auto row_group_index_reader = page_index_reader->RowGroup(i);
+      ASSERT_NE(row_group_index_reader, nullptr);
+
+      for (int j = 0; j < metadata->num_columns(); ++j) {
+        PageSizeStatistics page_stats;
+
+        auto column_index = row_group_index_reader->GetColumnIndex(j);
+        if (column_index != nullptr) {
+          if (column_index->has_repetition_level_histograms()) {
+            page_stats.ref_levels = column_index->repetition_level_histograms();
+          }
+          if (column_index->has_definition_level_histograms()) {
+            page_stats.def_levels = column_index->definition_level_histograms();
+          }
+        }
+
+        auto offset_index = row_group_index_reader->GetOffsetIndex(j);
+        if (offset_index != nullptr) {
+          page_stats.byte_array_bytes = offset_index->unencoded_byte_array_data_bytes();
+        }
+
+        page_stats_.emplace_back(std::move(page_stats));
+      }
+    }
+  }
+
+ protected:
+  std::shared_ptr<Buffer> buffer_;
+  std::vector<SizeStatistics> row_group_stats_;
+  std::vector<PageSizeStatistics> page_stats_;
+  inline static const SizeStatistics kEmptyRowGroupStats{};
+  inline static const PageSizeStatistics kEmptyPageStats{};
+};
+
+TEST_F(SizeStatisticsRoundTripTest, DisableSizeStats) {
+  auto schema = ::arrow::schema({
+      ::arrow::field("a", ::arrow::list(::arrow::list(::arrow::int32()))),
+      ::arrow::field("b", ::arrow::list(::arrow::list(::arrow::utf8()))),
+  });
+  WriteFile(SizeStatisticsLevel::None, ::arrow::TableFromJSON(schema, {R"([
+      [ [[1],[1,1],[1,1,1]], [["a"],["a","a"],["a","a","a"]] ],
+      [ [[0,1,null]],        [["foo","bar",null]]            ],
```

Comment:
So these 2 batches make one row group, and the 2 other batches below another row group, right? Can you add comments so that this is more immediately understood?